### ISSUE TITLE
Broken lambdas with C++ modules

### ISSUE THREAD

**Issue by vagran** [llvm/llvm-project#55634]
Consider the following code:

export module main;// <<< Remove this line to make it work correctly

import <iostream>;

export int
main()
{
    ([](){ std::cout << "callback 1\n"; })();
    ([](){ std::cout << "callback 2\n"; })();
    return 0;
}

It has the following output:
callback 1
callback 1

Expected output (it is correct, when the code compiled without modules):
callback 1
callback 2

I use the following commands for compiling:
CXX="/opt/clang-12/bin/clang++"
CXX_FLAGS="-std=c++20 -stdlib=libc++ -Wall -Werror -g -O0"
CXX_FLAGS="$CXX_FLAGS -fmodules -fimplicit-modules -fmodules-cache-path=modules-cache"
$CXX -c $CXX_FLAGS test.cpp -o test.o
$CXX $CXX_FLAGS test.o -o test_mod

clang version 14.0.0
Target: x86_64-unknown-linux-gnu

clang version 13.0.0 (https://github.com/llvm/llvm-project.git c1baf946e6cf611ae871e34db5cfea0f94f4b5a0)
Target: x86_64-unknown-linux-gnu

Also tested on: clang version 12.0.0 (https://github.com/llvm/llvm-project/ b978a93635b584db380274d7c8963c73989944a1)

Clang downloaded from GitHub release page.

I also created repository with reproduction code: https://github.com/vagran/asio-awaitable-modules-problem-demo/tree/exp-lambdas

I have reported this bug [here](https://bugs.llvm.org/show_bug.cgi?id=51607) year ago, but no any reaction since that time. It is actually show-stopper for using C++20 modules with Clang.

**Comment by JohelEGP** [comment]
This looks similar: https://github.com/llvm/llvm-project/issues/50455.

**Comment by vagran** [comment]
> This looks similar: #50455.

I do not think it is similar. Immediate invocation here is used to have minimal illustrative test case. The problem persist without immediate invocation as well (initially it was found when passing labdas to some ASIO functions). As I remember it might be related to some improper name mangling (some clash occurs) for lambdas when using C++20 modules.

**Comment by JohelEGP** [comment]
Yeah, not the surface, but the underlying issue.

**Comment by JohelEGP** [comment]
This may have been fixed by https://github.com/llvm/llvm-project/commit/e7eec38246560781e0a4020b19c7eb038a8c5655.

**Comment by vagran** [comment]
I tried to test it (built from latest `main`) with my test case but I cannot figure out how to include standard library files in new Clang (worked before, at least until Clang 14).
Current statements like `import <functional>` leads to this error:
```
../callback.cppm:6:8: error: header file <functional> (aka '/home/artyom/tmp/llvm-install/bin/../include/c++/v1/functional') cannot be imported because it is not known to be a header unit
```
Trying to include it in global module fragment like this:
```cpp
module;
#include <functional>

export module callback;
```
produces this error:
```
/home/artyom/tmp/llvm-install/bin/../include/x86_64-unknown-linux-gnu/c++/v1/module.modulemap:21:14: error: header 'ctype.h' not found
      header "ctype.h"
             ^
/home/artyom/tmp/llvm-install/bin/../include/c++/v1/stdint.h:123:15: note: submodule of top-level module 'std' implicitly imported here
#include_next <stdint.h>
```
Also tried to use modulemap.
`std_stub.h`:
```cpp
#include <iostream>
#include <functional>
```
`std_stub.modulemap`:
```
module std_stub {
    header "std_stub.h"
    export *
}
```
Compiling it with `$CXX $CXX_FLAGS -x c++-module -o std_stub.pcm ../std_stub.modulemap -fmodule-name=std_stub -Xclang -emit-module` emits the same error:
```
/home/artyom/tmp/llvm-install/bin/../include/x86_64-unknown-linux-gnu/c++/v1/module.modulemap:21:14: error: header 'ctype.h' not found
      header "ctype.h"
             ^
/home/artyom/tmp/llvm-install/bin/../include/c++/v1/stdint.h:123:15: note: submodule of top-level module 'std' implicitly imported here
#include_next <stdint.h>
              ^
```
Some hints would be appreciated.

**Comment by JohelEGP** [comment]
See https://github.com/llvm/llvm-project/issues/56012#issuecomment-1173958501.

**Comment by vagran** [comment]
> See [#56012 (comment)](https://github.com/llvm/llvm-project/issues/56012#issuecomment-1173958501).

Thanks, that helped. Tested the minimal test case in this issue as well as a bit more sophisticated from [here](https://github.com/vagran/asio-awaitable-modules-problem-demo/tree/separate-module-call). Both works correctly now.
